File: flutter/packages/aval_format/lib/src/graph_adapter.dart
/// Maps a validated compiled manifest into the canonical motion graph.
///
/// Dart port of `packages/format/src/graph-adapter.ts` (1.0). Depends on the
/// sibling `aval_graph` package for `validateMotionGraphDefinition`, which —
/// like the TypeScript original — accepts a genuinely untrusted, dynamically
/// shaped `Object?` value (a `Map`/`List` tree with the exact field names
/// `unitId`, `kind`, `frameCount`, `ports`, `portalFrames`, `entryFrame`,
/// `initialUnit`, `id`, `from`, `to`, `trigger`/`type`/`name`, `start`/`type`/
/// `sourcePort`/`targetPort`/`maxWaitFrames`, `transition`/`kind`/`unitId`/
/// `frameCount`/`direction`/`reverseOf`, `continuity`) rather than a typed
/// definition class, mirroring how `packages/graph/src/validate.ts` walks its
/// input as `unknown`.
library;

import 'package:aval_graph/aval_graph.dart'
    show MotionGraphValidationError, ValidatedMotionGraph, validateMotionGraphDefinition;

import 'errors.dart';
import 'model.dart';

/// Map a validated compiled manifest into the canonical motion graph.
ValidatedMotionGraph adaptManifestToMotionGraph(CompiledManifest manifest) {
  try {
    final unitsById = {for (final unit in manifest.units) unit.id: unit};
    final definition = <String, Object?>{
      'initialState': manifest.initialState,
      'states': manifest.states.map((state) => _adaptState(state, unitsById)).toList(),
      'edges': manifest.edges.map((edge) => _adaptEdge(edge, unitsById)).toList(),
    };
    return validateMotionGraphDefinition(definition);
  } on MotionGraphValidationError catch (error) {
    throw FormatError(
      FormatErrorCode.graphInvalid,
      'compiled manifest does not define a valid motion graph: ${error.message}',
    );
  } on FormatError catch (error) {
    if (error.code == FormatErrorCode.graphInvalid) rethrow;
    throw FormatError(
      FormatErrorCode.graphInvalid,
      'compiled manifest does not define a valid motion graph: ${error.message}',
    );
  } catch (error) {
    throw FormatError(
      FormatErrorCode.graphInvalid,
      'compiled manifest does not define a valid motion graph: $error',
    );
  }
}

Map<String, Object?> _adaptState(State state, Map<String, Unit> unitsById) {
  final body = unitsById[state.bodyUnit];
  if (body is! BodyUnit) {
    _graphInvalid('state ${_quote(state.id)} has no body unit');
  }
  final graphBody = <String, Object?>{
    'unitId': body.id,
    'kind': body.playback == 'loop' ? 'loop' : (body.frameCount == 1 ? 'held' : 'finite'),
    'frameCount': body.frameCount,
    'ports': body.ports
        .map((port) => <String, Object?>{
              'id': port.id,
              'entryFrame': 0,
              'portalFrames': [...port.portalFrames],
            })
        .toList(),
  };
  final base = <String, Object?>{'id': state.id, 'body': graphBody};
  if (state.initialUnit == null) {
    return base;
  }
  final initial = unitsById[state.initialUnit];
  if (initial is! OneShotUnit) {
    _graphInvalid('state ${_quote(state.id)} has no one-shot initial unit');
  }
  return {
    ...base,
    'initialUnit': {'unitId': initial.id, 'frameCount': initial.frameCount},
  };
}

Map<String, Object?> _adaptEdge(Edge edge, Map<String, Unit> unitsById) {
  final base = <String, Object?>{
    'id': edge.id,
    'from': edge.from,
    'to': edge.to,
    'start': _adaptStart(edge.start),
    'continuity': edge.continuity,
  };
  if (edge.trigger != null) {
    base['trigger'] = _adaptTrigger(edge.trigger!);
  }
  if (edge is NonCutEdge && edge.transition != null) {
    base['transition'] = _adaptTransition(edge.transition!, unitsById);
  }
  return base;
}

Map<String, Object?> _adaptTrigger(Trigger trigger) {
  if (trigger is EventTrigger) {
    return {'type': 'event', 'name': trigger.name};
  }
  return {'type': 'completion'};
}

Map<String, Object?> _adaptStart(Start start) {
  if (start is PortalStart) {
    return {
      'type': 'portal',
      'sourcePort': start.sourcePort,
      'targetPort': start.targetPort,
      'maxWaitFrames': start.maxWaitFrames,
    };
  }
  if (start is FinishStart) {
    return {'type': 'finish', 'targetPort': start.targetPort, 'maxWaitFrames': start.maxWaitFrames};
  }
  return {'type': 'cut', 'targetPort': start.targetPort, 'maxWaitFrames': 1};
}

Map<String, Object?> _adaptTransition(Transition transition, Map<String, Unit> unitsById) {
  final unit = unitsById[transition.unit];
  if (transition is LockedTransition) {
    if (unit is! BridgeUnit) {
      _graphInvalid('locked transition has no bridge unit ${_quote(transition.unit)}');
    }
    return {'kind': 'locked', 'unitId': unit.id, 'frameCount': unit.frameCount};
  }
  final reversible = transition as ReversibleTransition;
  if (unit is! ReversibleUnit) {
    _graphInvalid('reversible transition has no reversible unit ${_quote(transition.unit)}');
  }
  final base = <String, Object?>{
    'kind': 'reversible',
    'unitId': unit.id,
    'frameCount': unit.frameCount,
    'direction': reversible.direction,
  };
  return reversible.reverseOf == null ? base : {...base, 'reverseOf': reversible.reverseOf};
}

Never _graphInvalid(String message) {
  throw FormatError(FormatErrorCode.graphInvalid, message);
}

String _quote(String value) => '"$value"';

File: packages/format/src/graph-adapter.ts
import {
  validateMotionGraphDefinition,
  type GraphEdgeDefinition,
  type GraphStateDefinition,
  type GraphTransitionDefinition,
  type MotionGraphDefinition,
  type ValidatedMotionGraph
} from "@pixel-point/aval-graph";

import { FormatError } from "./errors.js";
import type {
  CompiledManifest,
  Edge,
  State,
  Transition,
  Unit
} from "./model.js";

/** Map a validated compiled manifest into the canonical motion graph. */
export function adaptManifestToMotionGraph(
  manifest: CompiledManifest
): ValidatedMotionGraph {
  try {
    const unitsById = new Map(manifest.units.map((unit) => [unit.id, unit]));
    const definition: MotionGraphDefinition = {
      initialState: manifest.initialState,
      states: manifest.states.map((state) => adaptState(state, unitsById)),
      edges: manifest.edges.map((edge) => adaptEdge(edge, unitsById))
    };
    return validateMotionGraphDefinition(definition);
  } catch (error) {
    if (error instanceof FormatError && error.code === "GRAPH_INVALID") {
      throw error;
    }
    const detail = error instanceof Error ? `: ${error.message}` : "";
    throw new FormatError(
      "GRAPH_INVALID",
      `compiled manifest does not define a valid motion graph${detail}`
    );
  }
}

function adaptState(
  state: State,
  unitsById: ReadonlyMap<string, Unit>
): GraphStateDefinition {
  const body = unitsById.get(state.bodyUnit);
  if (body?.kind !== "body") {
    graphInvalid(`state ${quote(state.id)} has no body unit`);
  }
  const graphBody = Object.freeze({
    unitId: body.id,
    kind:
      body.playback === "loop"
        ? "loop"
        : body.frameCount === 1
          ? "held"
          : "finite",
    frameCount: body.frameCount,
    ports: Object.freeze(
      body.ports.map((port) =>
        Object.freeze({
          id: port.id,
          entryFrame: 0 as const,
          portalFrames: Object.freeze([...port.portalFrames])
        })
      )
    )
  });
  const base = {
    id: state.id,
    body: graphBody
  };
  if (state.initialUnit === undefined) {
    return Object.freeze(base);
  }
  const initial = unitsById.get(state.initialUnit);
  if (initial?.kind !== "one-shot") {
    graphInvalid(`state ${quote(state.id)} has no one-shot initial unit`);
  }
  return Object.freeze({
    ...base,
    initialUnit: Object.freeze({
      unitId: initial.id,
      frameCount: initial.frameCount
    })
  });
}

function adaptEdge(
  edge: Edge,
  unitsById: ReadonlyMap<string, Unit>
): GraphEdgeDefinition {
  const trigger =
    edge.trigger === undefined ? undefined : Object.freeze({ ...edge.trigger });
  const start = Object.freeze({ ...edge.start });
  const transition =
    edge.transition === undefined
      ? undefined
      : adaptTransition(edge.transition, unitsById);
  const base = {
    id: edge.id,
    from: edge.from,
    to: edge.to,
    start,
    continuity: edge.continuity
  };
  if (trigger === undefined && transition === undefined) {
    return Object.freeze(base);
  }
  if (trigger === undefined) {
    return Object.freeze({ ...base, transition: transition! });
  }
  if (transition === undefined) {
    return Object.freeze({ ...base, trigger });
  }
  return Object.freeze({ ...base, trigger, transition });
}

function adaptTransition(
  transition: Transition,
  unitsById: ReadonlyMap<string, Unit>
): GraphTransitionDefinition {
  const unit = unitsById.get(transition.unit);
  if (transition.kind === "locked") {
    if (unit?.kind !== "bridge") {
      graphInvalid(`locked transition has no bridge unit ${quote(transition.unit)}`);
    }
    return Object.freeze({
      kind: "locked",
      unitId: unit.id,
      frameCount: unit.frameCount
    });
  }
  if (unit?.kind !== "reversible") {
    graphInvalid(
      `reversible transition has no reversible unit ${quote(transition.unit)}`
    );
  }
  const base = {
    kind: "reversible" as const,
    unitId: unit.id,
    frameCount: unit.frameCount,
    direction: transition.direction
  };
  return transition.reverseOf === undefined
    ? Object.freeze(base)
    : Object.freeze({ ...base, reverseOf: transition.reverseOf });
}

function graphInvalid(message: string): never {
  throw new FormatError("GRAPH_INVALID", message);
}

function quote(value: string): string {
  return JSON.stringify(value);
}
